### ISSUE TITLE
Any way to have a pre-defined List schema?

### ISSUE THREAD

**Issue by whoiswes** [marshmallow-code/marshmallow#1125]
Bad description.  Let me explain.  I have a legacy API that I'm inheriting, and unfortunately cannot rev the schema to be more sane.  I am attempting to parse the following JSON structure (part of a bigger object that's thankfully easy to map).  

Each of the key values is predefined, so in effect, we have 3 dictionaries with known keys in a list.

```
{
    "keys": [
        {"key_one": "blah1"},
        {"key_two": "blah2"},
        {"key_three": "blah3"}
    ]
}
```

The closest I've come is using a List and Dict type with a OneOf validator on the keys field, but that doesn't limit the input to exactly 3 dicts with all three keys represented.

```
class StupidSchema(Schema):
    """ Really dumb way of mapping a simple object"""
    keys = fields.List(
        fields.Dict(
            keys=fields.String(validate=OneOf(('key_one', 'key_two', 'key_three'))),
            values=fields.String(validate=Length(0, 100))
        )
    )
```

I'm also leveraging APISpec to generate my swagger def and the above doesn't give a clear indication of the allowed key values, which I can work around manually.

I've been looking at this for a few days now and am at a loss.  So throwing this out to the community to see if there's a cleaner way of accomplishing this.

Thanks for an awesome set of libraries!

**Comment by lafrech** [comment]
You could add a custom validator at list level to check the keys and number of elements.

Or you could use a custom field...

    keys = TripleBlahField()

That field would validate the content is a list as defined above with those three elements.

How you express that in apispec is another story. It could be related to https://github.com/marshmallow-code/apispec/issues/172.

**Comment by whoiswes** [comment]
Thanks @lafrech, I had looked at a custom field yesterday but gotten nowhere as it just wasn't clicking.  I'll have a look again and see if I can come up with something workable.

**Comment by deckar01** [comment]
I would squash that structure before loading to make it easier to use, then hard code the key constraints into a schema.

```py
from marshmallow import Schema, fields, pre_load
from marshmallow.validate import Length


class KeysSchema(Schema):
    key_one = fields.String(validate=Length(0, 100))
    key_two = fields.String(validate=Length(0, 100))
    key_three = fields.String(validate=Length(0, 100))

class LegacySchema(Schema):
    keys = fields.Nested(KeysSchema)
    
    @pre_load
    def flatten_keys(self, data):
        keys = data.get('keys', [])
        data['keys'] = {k: v for key in keys for k, v in key.items()}
        return data

data = {
    "keys": [
        {"key_one": "blah1"},
        {"key_two": "blah2"},
        {"key_three": "blah3"}
    ]
}
LegacySchema().load(data)
# {'keys': {'key_three': 'blah3', 'key_one': 'blah1', 'key_two': 'blah2'}}
```

If you need to dump the same structure you would have to do something similar with `@post_dump`.

**Comment by whoiswes** [comment]
@deckar01 - that's perfect.  Believe it or not my initial schema was set up very similar to the above, so I already have the post_dump code done.  I think I got hung up on the APISpec generation, but since there's not a native way of doing that anyway given the wonky structure, I've already manually created the schema definition and am simply overwriting the auto-generated one.

I just shifted things around and this works perfectly.  Just need to figure out why my swagger stuff isn't parsing correctly and I'll be golden.

@lafrech , @deckar01 - thank you both so much for your help!
